File: api/app/Modules/TravelAgency/Domain/Models/TravelBooking.php
<?php

declare(strict_types=1);

namespace App\Modules\TravelAgency\Domain\Models;

use App\Modules\TravelAgency\Domain\Enums\BookingSource;
use App\Modules\TravelAgency\Domain\Enums\BookingStatus;
use App\Modules\TravelAgency\Domain\Enums\PaymentStatus;
use App\Shared\Traits\BelongsToCompany;
use Database\Factories\TravelBookingFactory;
use Illuminate\Database\Eloquent\Builder;
use Illuminate\Database\Eloquent\Factories\HasFactory;
use Illuminate\Database\Eloquent\Model;
use Illuminate\Database\Eloquent\Relations\BelongsTo;
use Illuminate\Database\Eloquent\Relations\HasMany;
use Illuminate\Support\Carbon;
use Illuminate\Support\Str;

/**
 * Réservation multi-passagers (TRAVEL-209, issue #6022).
 *
 * `reference` est générée automatiquement (`GV-…`) si absente à la création.
 * `idempotency_key` garantit qu'une requête rejouée (retry réseau, double
 * clic guichet) ne crée jamais deux réservations pour le même tenant.
 */
/**
 * @property int $id
 * @property string $company_id
 * @property string $reference
 * @property int $trip_id
 * @property BookingStatus $status
 * @property int $passenger_count
 * @property int $total_amount_minor
 * @property string $currency
 * @property BookingSource $booking_source
 * @property int|null $customer_contact_id
 * @property int|null $booked_by_user_id
 * @property PaymentStatus $payment_status
 * @property Carbon|null $cancelled_at
 * @property string $cancel_reason
 * @property Carbon|null $expires_at
 * @property string $idempotency_key
 * @property int $version
 * @property string $contact_email
 * @property string $contact_phone
 * @property bool $notify_consent
 * @property Carbon|null $consent_recorded_at
 * @property string $round_trip_group_id
 * @property int $return_booking_id
 * @property string $leg
 * @property int $corporate_account_id
 * @property int $quote_id
 * @property bool $billing_deferred
 * @property string $connection_group_id
 * @property Carbon|null $created_at
 * @property Carbon|null $updated_at
 *
 * @mixin Builder<static>
 */
class TravelBooking extends Model
{
    use BelongsToCompany;

    /** @use HasFactory<TravelBookingFactory> */
    use HasFactory;

    protected $fillable = [
        'reference',
        'trip_id',
        'status',
        'passenger_count',
        'total_amount_minor',
        'currency',
        'booking_source',
        'customer_contact_id',
        'booked_by_user_id',
        'payment_status',
        'cancelled_at',
        'cancel_reason',
        'expires_at',
        'idempotency_key',
        'version',
        'contact_email',
        'contact_phone',
        'notify_consent',
        'consent_recorded_at',
        'round_trip_group_id',
        'return_booking_id',
        'leg',
        'corporate_account_id',
        'quote_id',
        'billing_deferred',
        'connection_group_id',
    ];

    protected $casts = [
        'passenger_count' => 'integer',
        'total_amount_minor' => 'integer',
        'status' => BookingStatus::class,
        'booking_source' => BookingSource::class,
        'payment_status' => PaymentStatus::class,
        'expires_at' => 'datetime',
        'cancelled_at' => 'datetime',
        'version' => 'integer',
        'notify_consent' => 'boolean',
        'consent_recorded_at' => 'datetime',
        'return_booking_id' => 'integer',
        'corporate_account_id' => 'integer',
        'quote_id' => 'integer',
        'billing_deferred' => 'boolean',
    ];

    protected static function booted(): void
    {
        static::creating(function (self $booking): void {
            if (empty($booking->reference)) {
                $booking->reference = self::generateReference();
            }

            if (empty($booking->idempotency_key)) {
                $booking->idempotency_key = (string) Str::uuid();
            }
        });
    }

    public static function generateReference(): string
    {
        return 'GV-'.strtoupper(Str::random(10));
    }

    /**
     * @return BelongsTo<TravelTrip, $this>
     */
    public function trip(): BelongsTo
    {
        return $this->belongsTo(TravelTrip::class, 'trip_id');
    }

    /**
     * @return HasMany<TravelPassenger, $this>
     */
    public function passengers(): HasMany
    {
        return $this->hasMany(TravelPassenger::class, 'booking_id');
    }

    /**
     * @return HasMany<TravelTicket, $this>
     */
    public function tickets(): HasMany
    {
        return $this->hasMany(TravelTicket::class, 'booking_id');
    }

    /**
     * @return HasMany<TravelPayment, $this>
     */
    public function payments(): HasMany
    {
        return $this->hasMany(TravelPayment::class, 'booking_id');
    }
}

File: front/mobile_apps/leopardo_travel_agent/lib/features/travel/models/travel_booking.dart
/// Réservation (TRAVEL-312..316, GET/POST /travel/bookings).
class TravelBooking {
  const TravelBooking({
    required this.id,
    this.reference,
    this.tripId,
    this.status,
    this.passengerCount,
    this.totalAmountMinor,
    this.currency,
    this.bookingSource,
    this.paymentStatus,
    this.expiresAt,
    this.contactEmail,
    this.contactPhone,
    this.notifyConsent,
    this.passengers = const [],
    this.tickets = const [],
  });

  final int id;
  final String? reference;
  final int? tripId;
  final String? status;
  final int? passengerCount;
  final int? totalAmountMinor;
  final String? currency;
  final String? bookingSource;
  final String? paymentStatus;
  final String? expiresAt;
  final String? contactEmail;
  final String? contactPhone;
  final bool? notifyConsent;
  final List<TravelPassenger> passengers;
  final List<TravelTicket> tickets;

  factory TravelBooking.fromJson(Map<String, dynamic> json) {
    return TravelBooking(
      id: json['id'] as int,
      reference: json['reference'] as String?,
      tripId: json['trip_id'] as int?,
      status: json['status'] as String?,
      passengerCount: json['passenger_count'] as int?,
      totalAmountMinor: json['total_amount_minor'] as int?,
      currency: json['currency'] as String?,
      bookingSource: json['booking_source'] as String?,
      paymentStatus: json['payment_status'] as String?,
      expiresAt: json['expires_at'] as String?,
      contactEmail: json['contact_email'] as String?,
      contactPhone: json['contact_phone'] as String?,
      notifyConsent: json['notify_consent'] as bool?,
      passengers: (json['passengers'] as List?)
              ?.whereType<Map>()
              .map((e) => TravelPassenger.fromJson(e.cast<String, dynamic>()))
              .toList() ??
          const [],
      tickets: (json['tickets'] as List?)
              ?.whereType<Map>()
              .map((e) => TravelTicket.fromJson(e.cast<String, dynamic>()))
              .toList() ??
          const [],
    );
  }
}

/// Passager d'une réservation (RGPD : jamais le n° de pièce d'identité).
class TravelPassenger {
  const TravelPassenger({
    this.id,
    this.bookingId,
    this.fullName,
    this.birthDate,
    this.documentType,
    this.hasDocument = false,
    this.ageCategory,
    this.classId,
    this.seatNumber,
    this.unitPriceMinor,
  });

  final int? id;
  final int? bookingId;
  final String? fullName;
  final String? birthDate;
  final String? documentType;
  final bool hasDocument;
  final String? ageCategory;
  final int? classId;
  final int? seatNumber;
  final int? unitPriceMinor;

  factory TravelPassenger.fromJson(Map<String, dynamic> json) {
    return TravelPassenger(
      id: json['id'] as int?,
      bookingId: json['booking_id'] as int?,
      fullName: json['full_name'] as String?,
      birthDate: json['birth_date'] as String?,
      documentType: json['document_type'] as String?,
      hasDocument: json['has_document'] as bool? ?? false,
      ageCategory: json['age_category'] as String?,
      classId: json['class_id'] as int?,
      seatNumber: json['seat_number'] as int?,
      unitPriceMinor: json['unit_price_minor'] as int?,
    );
  }
}

/// Billet émis (TRAVEL-316..317, POST /travel/bookings/{id}/issue-ticket).
class TravelTicket {
  const TravelTicket({
    this.id,
    this.ticketNumber,
    this.bookingId,
    this.passengerId,
    this.status,
    this.issuedAt,
    this.validFrom,
    this.validUntil,
    this.checkedInAt,
  });

  final int? id;
  final String? ticketNumber;
  final int? bookingId;
  final int? passengerId;
  final String? status;
  final String? issuedAt;
  final String? validFrom;
  final String? validUntil;
  final String? checkedInAt;

  factory TravelTicket.fromJson(Map<String, dynamic> json) {
    return TravelTicket(
      id: json['id'] as int?,
      ticketNumber: json['ticket_number'] as String?,
      bookingId: json['booking_id'] as int?,
      passengerId: json['passenger_id'] as int?,
      status: json['status'] as String?,
      issuedAt: json['issued_at'] as String?,
      validFrom: json['valid_from'] as String?,
      validUntil: json['valid_until'] as String?,
      checkedInAt: json['checked_in_at'] as String?,
    );
  }
}
